flax/linen/attention.py

```diff
@@ -33,147 +33,99 @@
 
 PRNGKey = Any
 Shape = Tuple[int]
-Dtype = Any  # this could be a real type?
+Dtype = Any
 Array = Any
 
 
-def dot_product_attention(query,
-                          key,
-                          value,
-                          dtype=jnp.float32,
-                          bias=None,
-                          axis=None,
-                          broadcast_dropout=True,
-                          dropout_rng=None,
-                          dropout_rate=0.,
-                          deterministic=False,
-                          precision=None):
+def dot_product_attention(query: Array,
+                          key: Array,
+                          value: Array,
+                          bias: Optional[Array] = None,
+                          broadcast_dropout: bool = True,
```

Comment:
Do you think we should get rid of the "inlined dropout" too?
One way is to make the attention function dynamic like in the design tests.

Reply:
I agree that we should decouple it more.  it's already a dynamic attribute but the expected signature of the Callable is too specialized, it should just take q,k,v,bias.  I think I'd rather do that change in a separate PR though.